rust/Dockerfile

```diff
@@ -1,19 +1,23 @@
-FROM rust:1.72-slim as BUILDER
+FROM lukemathwalker/cargo-chef:latest-rust-1.72-slim-bookworm as chef
+WORKDIR /build
+
+FROM chef as planner
+COPY . .
 ARG PACKAGE
-WORKDIR /build/
-COPY . ./
-RUN --mount=type=cache,target=./target \
-    --mount=type=cache,target=/usr/local/cargo/registry,sharing=locked \
-    --mount=type=cache,target=/usr/local/rustup,sharing=locked \
-    cargo build -p $PACKAGE --release
+RUN cargo chef prepare --recipe-path recipe.json
 
-RUN --mount=type=cache,target=./target \
-    mv ./target/release/$PACKAGE /usr/local/bin/$PACKAGE
+FROM chef as builder
+COPY --from=planner /build/recipe.json recipe.json
+RUN cargo chef cook --release --recipe-path recipe.json
```

Comment:
This essentially triggers the layer caching.

Reply:
Not sure how it is supposed to work if we don't `COPY` the toml file with the deps beforehand? How it will learn which deps to fetch and compile?

Reply:
I tested this config with cloud build and layer cache doesn't work with this dockerfile

Reply:
Yeah, looks like layer caching is not working for GitHub actions either (for Elixir too):

https://github.com/firezone/firezone/actions/runs/6399364604/job/17371369030?pr=2233#step:9:292

Going to spend a few minutes making sure we're good there.

`cargo chef` handles the dependency caching for us:

> Nothing too mysterious going on here, you can examine the recipe.json file: it contains the skeleton of your project (e.g. all the Cargo.toml files with their relative path, the Cargo.lock file is available) plus a few additional pieces of information.
In particular it makes sure that all libraries and binaries are explicitly declared in their respective Cargo.toml files even if they can be found at the canonical default location (src/main.rs for a binary, src/lib.rs for a library).

> The recipe.json is the equivalent of the Python requirements.txt file - it is the only input required for cargo chef cook, the command that will build out our dependencies:

Reply:
@jamilbk It can not work, look at this stage:
```
COPY . .
ARG PACKAGE
RUN cargo chef prepare --recipe-path recipe.json
```
`COPY . .` will reset any cache on any file change, everything after that will be rebuilt. Lemme try to fix it in my branch to make the cache work.

Reply:
> COPY . . will reset any cache on any file change,

Correct. But this is in a different build stage. `recipe.json` is generated with the dependency changes only which gets passed to the next build stage whose cache should not be busted by the `COPY . .` in the stage before.

Reply:
Rust caching is working locally -- changing a source file does not bust the dependency cache. The issue in CI is with the `type=gha` cache specifically with buildx I suspect.

Reply:
Ah, so I experimented a bit more and the trick is that if recipe.json is unchanged it's the other staging which will have the same checksum and continue the layer cache, it is used in Cloud Build:
```
#11 DONE 35.0s

#12 [builder 1/5] COPY --from=planner /build/recipe.json recipe.json
#12 CACHED

#13 [builder 2/5] RUN cargo chef cook --release --recipe-path recipe.json
#13 CACHED
```